design/code-asset-metadata.md

```diff
@@ -0,0 +1,46 @@
+# RFC: AWS Lambda - Metadata about Code Assets
+
+As described in [#1432](https://github.com/awslabs/aws-cdk/issues/1432), in order to support local debugging,
+debuggers like [SAM CLI](https://github.com/awslabs/aws-sam-cli) need to be able to find the code of a Lambda
+function locally.
+
+The current implementation of assets uses CloudFormation Parameters which represent the S3 bucket and key of the
+uploaded asset, which makes it impossible for local tools to reason about (without traversing the cx protocol with
+many heuristics).
+
+## Approach
+
+We will automatically embed CloudFormation metadata on `AWS::Lambda::Function` resources which use
```

Comment:
Why only `AWS::Lambda::Function`s? What about `AWS::Lambda::LayerVersion`, `AWS::Serverless::Function`, and `AWS::Serverless::LayerVersion`?

We are purposefully excluding API Gateway from this correct?

Is there a generic way this could work to allow any resource using an Asset to get this metadata? Trying to think broader here to make less work in the future of manually adding this metadata to every resource.

Reply:
The implementation allows anyone who uses an asset to attach this metadata to the resource by calling `asset.addResourceMetadata(resource, prop)`. I decided against doing this automatically because this problem is basically confined to only L2 constructs. Any higher level constructs shouldn't care about this at all, so I opted for a more manual solution.

Layers are currently in PR (#1411) and @RomainMuller should probably.

Reply:
@eladb What about `AWS::Serverless::Function`?

Reply:
@jfuss the `AWS::Serverless::Function` is currently only supported as an L1 construct (`serverless.CfnFunction`), and I suspect people who use it will likely not try to use it with CDK assets but rather with `CodeUri`. For example:

```ts
new serverless.CfnFunction(this, 'Func', {
  codeUri: 'file://mycode',
   runtime: 'nodejs8.10',
   handler: 'index.handler'
});
```

If users wish to use CDK assets (and invoke them locally through SAM CLI), this is what they will have to do:

```ts
const asset = new assets.ZipDirectoryAsset(this, 'Foo', {
  path: '/foo/boom'
});

const resource = new serverless.CfnFunction(this, 'Func', {
  codeUri: {
    bucket: asset.s3BucketName,
    key: asset.s3ObjectKey
  },
  runtime: 'nodejs8.10',
  handler: 'index.handler'
});

resource.addResourceMetadata(resource, 'CodeUri');
```

But TBH, I doubt that this makes sense. If you are already writing "CDK native" code, the L2 constructs for Lambda, API Gateway, etc provides a much richer API then the `AWS::Serverless::Function` resource.

What do you think?

Reply:
@eladb I see. So it's more like `Assets` are directly supported within `AWS::Serverless::Function` like it is with `AWS::Lambda::Function`s (over simplifying here due to the Construct level difference). I see Assets being powerful and something we should consider supporting directly but out of scope for this PR. 

Thanks for clarifying this.